http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AzurePubsubHandler.scala

```diff
@@ -697,140 +823,155 @@ class AzurePubsubHandlerInterp[F[_]: Parallel](
-                    s"runtime ${msg.runtimeId} is deleted successfully, moving to disk deletion"
-                  )
-                  _ <- deleteDiskAction
-                  _ <- dbRef.inTransaction(clusterQuery.completeDeletion(runtime.id, ctx.now))
-                  _ <- logger.info(ctx.loggingCtx)(
-                    s"runtime ${msg.runtimeId} ${if (msg.diskIdToDelete.isDefined) "and associated disk"} have been deleted successfully"
-                  )
-                } yield ()
-              case WsmJobStatus.Failed =>
-                F.raiseError[Unit](
-                  AzureRuntimeDeletionError(
-                    msg.runtimeId,
-                    msg.workspaceId,
-                    s"WSM delete VM job failed due to ${resp.errorReport.map(_.message).getOrElse("unknown")}"
-                  )
-                )
-              case WsmJobStatus.Running =>
-                F.raiseError[Unit](
-                  AzureRuntimeDeletionError(
-                    msg.runtimeId,
-                    msg.workspaceId,
-                    s"WSM delete VM job was not completed within ${config.deleteVmPollConfig.maxAttempts} attempts with ${config.deleteVmPollConfig.interval} delay"
-                  )
-                )
-            }
-          } yield ()
+      // Even if there is no wsm record yet, we may still have a leo disk record to clean up
+      leoDiskCleanupActionOpt = msg.diskIdToDelete.traverse { diskId =>
+        dbRef.inTransaction(persistentDiskQuery.delete(diskId, ctx.now))
+      }
 
-          _ <- asyncTasks.offer(
-            Task(
-              ctx.traceId,
-              taskToRun,
-              Some { e =>
-                handleAzureRuntimeDeletionError(
-                  AzureRuntimeDeletionError(msg.runtimeId,
-                                            msg.workspaceId,
-                                            s"Fail to delete runtime due to ${e.getMessage}"
-                  )
-                )
-              },
-              ctx.now,
-              "deleteAzureRuntime"
-            )
+      // Formulate the WSM call and polling params for later use in async VM deletion
+      deleteVmActionOpt = msg.wsmResourceId.fold {
+        // Error'd runtimes might not have a WSM resourceId. We still want deletion to succeed in this case.
+        logger
+          .info(ctx.loggingCtx)(
+            s"No VM wsmResourceId found for delete azure runtime msg $msg. No-op for wsmApi.deleteAzureVm."
           )
-        } yield ()
+          .map(_ => none[PollVmParams])
+      } { resourceId =>
+        for {
+          deleteJobControl <- F.delay(getDeleteControlledResourceRequest)
+          _ <- F.delay(wsmApi.deleteAzureVm(deleteJobControl, msg.workspaceId.value, resourceId.value))
+        } yield Some(PollVmParams(msg.workspaceId, WsmJobId(deleteJobControl.getJobControl.getId), runtime))
       }
+
+      // Delete and poll on all associated resources in WSM, and then mark the runtime as deleted
```

Comment:
I think it would be good to explicitly list out the resource types here:
- Hybrid connection (handled by leo directly not wsm)
- Storage container
- Disk
- VM
- other?